File: node/cms.js
// Get CMS data from external REST API and store it in a file
const fs = require('fs')
const axios = require('axios').default
const CMS_URL = 'http://jsonplaceholder.typicode.com/posts'

// Get posts, returns an array of post objects
// Responsible for everything around the CMS DB, including:
// transport rotocol (HTTP), location (URI) and payload schema (JSON)
async function cms_get_posts(fetch) {
  try {
    var result = await fetch.get(CMS_URL)
    console.log('getting CMS data')
  } catch (err) {
    console.log('an error occurred getting the post data', err)
    return Promise.reject('error')
  }
  if (result.status === 200) {
    return Promise.resolve(result.data)
  } else {
    console.log('error occurred, status =', result.status)
    return Promise.reject('error')
  }
}

function write_report(fs, report) {
  const today = new Date()
  const today_date = today.toISOString().slice(0,10)
  const filename = 'cms-' + today_date + '.json'
  fs.writeFileSync(filename, JSON.stringify(report))
}

module.exports = {
  main: async function(context, filesystem = fs, fetch = axios) {
    const posts = await cms_get_posts(fetch)
    console.log('CMS data has: ', posts.length, 'records')

    let summary = { posts: 0, users: 0, posts_per_user: 0 }
    let users = []
    for (let i = 0; i < posts.length; i++) {
      // console.log('processing record', i, 'for user', res.data[i].userId)
      summary.posts++
      if (!users.includes(posts[i].userId)) {
        users.push(posts[i].userId)
      }
    }
    summary.users = users.length
    summary.mean_posts_per_user = Math.round(summary.posts / summary.users)

    write_report(filesystem, summary)
    console.log('wrote CMS report')

    return Promise.resolve(true)
  }
}

File: java/src/main/java/CMS.java
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.net.URL;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.time.format.DateTimeFormatter.ISO_DATE;

public class CMS {
    private static final String REPORT_DIRECTORY = System.getProperty("java.io.tmpdir");
    private static final String REPORT_NAME = String.format("cms-%s.json", LocalDateTime.now().format(ISO_DATE));
    private static final File REPORT_FILE = new File(REPORT_DIRECTORY, REPORT_NAME);

    public void storeCmsSummaryReport() {
        System.out.println("Getting CMS data");
        String postsText = "";
        try {
            URL postsUrl = new URL("https://jsonplaceholder.typicode.com/posts");
            try (Scanner scanner = new Scanner(postsUrl.openStream(), UTF_8)) {
                scanner.useDelimiter("\\A");
                postsText = scanner.hasNext() ? scanner.next() : "";
            }
        } catch (IOException ex) {
            System.err.printf("No valid JSON response from CMS, error %s", ex.getMessage());
        }

        JSONArray posts = new JSONArray(postsText);
        System.out.printf("CMS data has %d records%n", posts.length());

        Set<Integer> userIds = new HashSet<>();
        posts.forEach(post -> userIds.add(((JSONObject)post).getInt("userId")));

        try(BufferedWriter writer = new BufferedWriter(new FileWriter(REPORT_FILE))) {
            int postCount = posts.length();
            int userCount = userIds.size();
            String summary = String.format("{ \"posts\": %d, \"users\": %d, \"mean_posts_per_user\": %d }",
                    postCount, userCount, Math.round((float)postCount / (float)userCount));
            writer.write(summary);
        } catch (IOException e) {
            System.err.printf("Unable to write report %s: %s%n",
                    REPORT_FILE.getAbsolutePath(), e.getMessage());
        }
        System.out.println("Wrote CMS report");
    }

    public static void main(String[] args) {
        new CMS().storeCmsSummaryReport();
    }
}
